listenbrainz/webserver/static/js/src/UserArtists.tsx

```diff
@@ -0,0 +1,192 @@
+
+export type UserArtistsData = Array<{
+  id: string;
+  count: number;
+}>;
+
+export type UserArtistsProps = {
+  user: ListenBrainzUser;
+  apiUrl: string;
+};
+
+export type UserArtistsState = {
+  data: UserArtistsData;
+};
+
+export default class UserArtists extends React.Component<
+  UserArtistsProps,
+  UserArtistsState
+> {
+  APIService: APIService;
+
+  private ROWS_PER_PAGE = 25; // Number of atists to be shown on each page
+  private maxListens = 0; // Number of listens for first artist used to scale the graph
+  private totalPages = 0; // Total namber of pages, useful for pagination
+  private currPage = 1; // Current page number
+
+  constructor(props: UserArtistsProps) {
+    super(props);
+
+    this.APIService = new APIService(
+      props.apiUrl || `${window.location.origin}/1`
+    ); // Used to access LB API
+
+    this.state = {
+      data: [],
+    };
+  }
+
+  async componentDidMount() {
+    const { user } = this.props;
+
+    // Fetch page number from URL
+    const url = new URL(window.location.href);
+    if (url.searchParams.get("page")) {
+      this.currPage = Number(url.searchParams.get("page"));
+    } else {
+      this.currPage = 1;
+    }
+
+    try {
+      const data = await this.APIService.getUserStats(
+        user.name,
+        undefined,
+        undefined,
+        1
+      );
+      this.maxListens = data.payload.artists[0].listen_count;
+      this.totalPages = Math.ceil(
+        data.payload.total_artist_count / this.ROWS_PER_PAGE
+      );
+    } catch (error) {
+      this.handleError(error);
+    }
+    this.handlePageChange(this.currPage);
+  }
+
+  getData = async (
+    userName: string,
+    offset: number
+  ): Promise<UserArtistsResponse> => {
+    const data = await this.APIService.getUserStats(
+      userName,
+      undefined,
+      offset,
+      this.ROWS_PER_PAGE
+    );
+    return data;
+  };
+
+  processData = (
+    data: UserArtistsResponse,
+    offset: number
+  ): UserArtistsData => {
+    const result = data.payload.artists
+      .map((elem, idx: number) => {
+        return {
+          id: `${offset + idx + 1}. ${elem.artist_name}`,
+          count: elem.listen_count,
+        };
+      })
+      .reverse();
+    return result;
+  };
+
+  handlePageChange = async (page: number): Promise<void> => {
+    const { user } = this.props;
+    const offset = (page - 1) * this.ROWS_PER_PAGE;
+    try {
+      const data = await this.getData(user.name, offset);
+      this.currPage = page;
+      this.setState({ data: this.processData(data, offset) });
+    } catch (error) {
+      this.handleError(error);
+    }
+  };
+
+  handleError = (error: Error): void => {
+    // Error Boundaries don't catch errors in async code.
+    // Throwing an error in setState fixes this.
+    // This is a hacky solution but should be fixed with upcoming concurrent mode in React.
+    const { user } = this.props;
+
+    if (
+      error.name === "SyntaxError" &&
```

Comment:
could we check the http code instead somehow, this seems risky and something that could break easily?